rllib/agents/dqn/learner_thread.py

```diff
@@ -54,22 +84,115 @@ def step(self):
                         td_error = results["td_error"]
                         # Switch off auto-conversion from numpy to torch/tf
                         # tensors for the indices. This may lead to errors
                         # when sent to the buffer for processing
                         # (may get manipulated if they are part of a tensor).
-                        replay.policy_batches[pid].set_get_interceptor(None)
+                        ma_batch.policy_batches[pid].set_get_interceptor(None)
                         prio_dict[pid] = (
-                            replay.policy_batches[pid].get("batch_indexes"),
+                            ma_batch.policy_batches[pid].get("batch_indexes"),
                             td_error,
                         )
                     self.learner_info = learner_info_builder.finalize()
-                    self.grad_timer.push_units_processed(replay.count)
-                self.outqueue.put((ra, prio_dict, replay.count))
+                    self.grad_timer.push_units_processed(ma_batch.count)
+                self.outqueue.put((replay_actor, prio_dict, ma_batch.count))
+                self.weights_updated = True
+
             self.learner_queue_size.push(self.inqueue.qsize())
+            self.overall_timer.push_units_processed(ma_batch and ma_batch.count or 0)
+
+
+# For backward compatibility.
+LearnerThread = APEXLearnerThread
+
+
+class APEXMultiGPULearnerThread(APEXLearnerThread):
+    """Background thread that updates the local model from replay data.
+
+    Thereby utilizing multiple GPUs via pre data loading, then updating
+    from the pre-loaded data.
+    """
+
+    def __init__(
+        self,
+        local_worker: RolloutWorker,
+        train_batch_size: int = 500,
+        num_multi_gpu_tower_stacks: int = 1,
+        num_data_load_threads: int = 16,
+    ):
+        """Initializes a MultiGPULearnerThread instance.
+
+        Args:
+            local_worker (RolloutWorker): Local RolloutWorker holding
+                policies this thread will call load_data() and optimizer() on.
+            train_batch_size (int): Size of batches (minibatches if
+                `num_sgd_iter` > 1) to learn on.
+            num_multi_gpu_tower_stacks (int): Number of buffers to parallelly
+                load data into on one device. Each buffer is of size
+                `train_batch_size` and hence increases GPU memory usage
+                accordingly.
+            num_data_load_threads (int): Number of threads to use to load
+                data into GPU memory in parallel.
+        """
+        super().__init__(local_worker=local_worker)
+
+        self.train_batch_size = train_batch_size
+
+        # TODO: (sven) Allow multi-GPU to work for multi-agent as well.
+        self.policy = self.local_worker.policy_map[DEFAULT_POLICY_ID]
+        self.policy_map = self.local_worker.policy_map
+        self.devices = self.policy.devices
```

Comment:
curious, where do we verify that there are that many physical GPU devices for batch loading and learning?
does it happen in policy?

Reply:
Yes, in the Policy c'tor. There, we check the config as well as the ray assigned devices and then check for their validity using tf and torch tools.